### PR TITLE
Moved MXNet implementation from Sound Classifier into a new class

### PR DESCRIPTION
**Pull request by fareeha**
Removed MXNet impl from Sound Classifier `create()` and `CustomModel` class into an MXNet specific class `MultiLayerPerceptronMXNetModel`. The changes also support save and load from old impl to the new one. 

Remaining MXNet not replaced yet are: MXNet data iterator and metric calculator.

### CODE REVIEW

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import mxnet as mx
+import time
+import turicreate as _tc
+
+from .._mxnet import _mxnet_utils
+
+class MultiLayerPerceptronMXNetModel():
+
+
+	def __init__(self, feature_output_length, num_labels, custom_layer_sizes, verbose):
```
Comment (nickjong) [inline src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: Please don't use tabs for indentation. Let's follow the conventions elsewhere in our codebase, and use four spaces for indentation in Python code

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		self.ctx = _mxnet_utils.get_mxnet_context()
+		self.verbose = verbose
+		self.custom_NN = self._build_custom_neural_network(feature_output_length, num_labels, custom_layer_sizes)
+		self.custom_NN.initialize(mx.init.Xavier(), ctx=self.ctx)
+
+		self.trainer = mx.gluon.Trainer(self.custom_NN.collect_params(), 'nag', {'learning_rate': 0.01, 'momentum': 0.9})
+		self.softmax_cross_entropy_loss = mx.gluon.loss.SoftmaxCrossEntropyLoss()
+
+
+	def train(self, data, label):
```
Comment (nickjong) [inline src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: Could we document the expected types and shapes of these arguments? And for the other methods, too?

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		# Inside training scope
+		data_shape = data.shape[0]
```
Comment (nickjong) [inline src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: What is a typical value of `data_shape`? Isn't this really just the batch size?

File: src/python/turicreate/toolkits/sound_classifier/sound_classifier.py [hunk 4]
```diff
@@ -289,7 +273,7 @@ def create(dataset, target, feature, max_iterations=10,
 
     state = {
         '_class_label_to_id': class_label_to_id,
-        '_custom_classifier': custom_NN,
+        '_custom_classifier': sc_model,
```
Comment (TobyRoseman) [inline src/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: I would prefer to keep the variable name `custom_NN`. I think that's a more descriptive name. It also matches language used in the user guide.

Comment (fareeha) [reply src/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: We refer to our Sequential model as `custom_NN` in the MXNet class so I didn't want to make it confusing. Any thoughts @nickjong ?

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py (hunk 3 above)
Comment (fareeha) [reply src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: It can be <=`batch_size` depending on the size of data passed. If input data is less than batch size, `data_shape` would have a value less than batch size.

File: src/python/turicreate/toolkits/sound_classifier/sound_classifier.py (hunk 4 above)
Comment (nickjong) [reply src/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: I think `custom_NN` probably is the clearer name, since this object only contains the neural network classifier, not the audio preprocessing and feature extraction that are also part of the overall Sound Classifier model

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        with mx.autograd.record():
+            for x, y in zip(data, label):
+                z = self.custom_NN(x)
+                # Computes softmax cross entropy loss.
+                loss = self.softmax_cross_entropy_loss(z, y)
+                # Backpropagate the error for one iteration.
+                loss.backward()
+        # Make one step of parameter update. Trainer needs to know the
+        # batch size of data to normalize the gradient by 1/batch_size.
+        self.trainer.step(data_shape)
```
Comment (nickjong) [inline src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: This comment describes the information being passed to `self.trainer.step` as the batch size. Why not just call the variable `batch_size`, since by construction we know it is a scalar value, the first element of the data's shape?

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+        self.trainer = mx.gluon.Trainer(self.custom_NN.collect_params(), 'nag', {'learning_rate': 0.01, 'momentum': 0.9})
+        self.softmax_cross_entropy_loss = mx.gluon.loss.SoftmaxCrossEntropyLoss()
+
+
+    def train(self, data, label):
+        """
+        Parameters
+        ----------
+        data : NumPy Array
```
Comment (nickjong) [inline src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: Could you document the meaning of the parameter, not just the expected type?

File: src/python/turicreate/toolkits/sound_classifier/sound_classifier.py [hunk 7]
```diff
@@ -825,11 +788,8 @@ class as a vector. Label ordering is dictated by the ``classes``
             if batch.pad != 0:
                 batch_data = batch_data[:-batch.pad]    # prevent batches looping back
 
-            batch_data = mx.gluon.utils.split_and_load(batch_data, ctx_list=ctx, batch_axis=0, even_split=False)
-
-            for x in batch_data:
-                forward_output = self._custom_classifier.forward(x)
-                y += mx.nd.softmax(forward_output).asnumpy().tolist()
+            forward_output = self._custom_classifier.predict(batch_data)
+            y += mx.nd.softmax(mx.nd.array(forward_output[0])).asnumpy().tolist()
```
Comment (nickjong) [inline src/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: This leaves some MXNet code outside of the MXNet encapsulation. Could we move the application of `softmax` inside `predict`?

File: src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py (hunk 5 above)
Comment (fareeha) [reply src/python/turicreate/toolkits/sound_classifier/_mx_sound_classifier.py]: Sure I can make it `batch_size = min(batch_size, data.shape[0])` since `data.shape[0]` can be smaller than batch_size if examples it holds are fewer.
